[execute_process] Don't strip embedded semicolons

jgehw opened the pull request:
The existing implementation totally screwed up commands if the command's arguments contained semicolons (this is the case, e.g., in the FindPython modules of the cmake distribution).

Fixes #12922.

Commit:
fix execute process macro
The existing implementation totally screwed up commands if the command's arguments contained semicolons (this is the case, e.g., in the FindPython modules of the cmake distribution).

Commit:
extend overriden execute_process to more than one COMMAND as there actually are use cases for this

jgehw commented:
Added support for execute_process with multiple COMMANDs as there actually are use cases in both vcpkg itself and some ports.

jgehw committed:
Merge pull request #1 from microsoft/master

incorporate changes from microsoft:master

jgehw committed:
Revert "incorporate changes from microsoft:master"

jgehw committed:
Merge pull request #2 from jgehw/revert-1-master

Revert "incorporate changes from microsoft:master"

jgehw committed:
Revert "Revert "incorporate changes from microsoft:master""

Commit:
Merge branch 'master' of https://github.com/microsoft/vcpkg

Commit:
Merge branch 'master' of https://github.com/microsoft/vcpkg

BillyONeal commented on scripts/cmake/execute_process.cmake:
```diff
@@ -13,8 +13,75 @@ if (NOT DEFINED OVERRIDEN_EXECUTE_PROCESS)
       message(FATAL_ERROR "This command cannot be executed in Download Mode.\nHalting portfile execution.\n")
     endmacro()
   else()
```
In this `else` case could we just do nothing instead?

Commit:
Merge branch 'master' of https://github.com/microsoft/vcpkg

Commit:
Merge branch 'master' of https://github.com/microsoft/vcpkg

jgehw commented:
Why did you close this PR? #12977 is no fix for this case. We either need this fix or [ras0219@b76640e].

BillyONeal commented:
@jgehw I didn't actively try to close this, GitHub does that itself if a PR is merged that mentions another issue or PR with a hot phrase like "fixes 1234" in the discussion / history.

BillyONeal commented:
(Specifically that PR had "fixed #12926" in the description which causes GitHub to link them together and close one when the other is merged)

jgehw commented:
@BillyONeal Ahh, I see, I'll pay attention to avoid such phrasing next time.

jgehw committed:
fix command (it may contain spaces such as C:\Program Files\...)

jgehw committed:
revert last commit (root cause for CI failures is something different: there are line breaks in path)

jgehw committed:
fix regex comparision
(value needs to be escaped as it may contains regex special characters such as brackets, eg C:/Program Files (x86)/...)

jgehw commented:
Found and fixed a bug in regex matching, which occurred in CI testing.

JackBoosY commented:
`execute_process` is often used in vcpkg, we must carefully modify this function.

jgehw commented:
> `execute_process` is often used in vcpkg, we must carefully modify this function.

- I implemented it with care, always knowing what I'm doing.
- Nevertheless I made a mistake (forgot to escape regex special chars before matching, see fix in last commit). This mistake was detected in CI testing.
- Beside CI testing, I suggest to have three independent code reviewers, not just one.

Do think this should suffice? If not, do you have any more ideas how to add additional layers of care?

BillyONeal commented:
I would like to see Robert's version that doesn't keep us needing to constantly sync with the upstream definition of execute_process.

jgehw committed:
Merge branch 'master' of https://github.com/microsoft/vcpkg

jgehw committed:
replace solution by easy-to-maintain work-around as discussed in PR
that fixes execute_process and only leaves _execute_process behind being broken

jgehw commented:
ok, I replaced my solution with Robert's implementation

BillyONeal commented:
@ras0219 OK we tried your workaround but it causes stuff to break.

Maybe we should change all the `_execute_process` calls to `vcpkg_execute_in_download_mode` or similar?

jgehw committed:
introduced vcpkg_execute_in_download_mode() and replaced _execute_process() by it

jgehw commented:
> Maybe we should change all the `_execute_process` calls to `vcpkg_execute_in_download_mode` or similar?

@BillyONeal Did this as you suggested. I'm curious for the CI results ;-)

jgehw committed:
Merge branch 'master' into fix-execute-process-macro2

jgehw committed:
add missing includes

jgehw committed:
Merge branch 'master' of https://github.com/microsoft/vcpkg

jgehw committed:
Merge branch 'master' into 'fix-execute-process-macro2'

jgehw commented:
I don't see a link between last nights build errors and this PR's code changes. Some errors sound transient, e.g.
`-- The default mirror is in Disaster Recovery mode, trying other mirrors...`; others such as unresolved symbols during linking sound unrelated -- merging master branch to ensure we're not missing any fixes here.

jgehw commented:
The errors are gone now except for the selene port, but I saw this also failing in other branches, thus most likely an issue with selene.

jgehw committed:
Merge branch 'master' of https://github.com/microsoft/vcpkg

jgehw committed:
Merge branch 'master' into fix-execute-process-macro2

jgehw committed:
Merge branch 'master' into fix-execute-process-macro2

jgehw committed:
remove unnecessary include after merge of master

jgehw committed:
Merge branch 'master' of https://github.com/microsoft/vcpkg

jgehw committed:
Merge branch 'master' of https://github.com/microsoft/vcpkg into master

jgehw committed:
Merge branch 'master' into fix-execute-process-macro2

jgehw commented:
arm_uwp and x64_uwp CI failing on json5-parser port downloading https://bitbucket.com/wlandry/json5_parser/get/1.0.0.tar.gz. When viewing in a browser, I get "Repository wlandry/json5_parser not found". So, not related to this PR.

BillyONeal commented:
Yeah, upstream is gone.

BillyONeal commented:
/azp run

jgehw commented:
/azp run

jgehw committed:
Merge branch 'master' of https://github.com/microsoft/vcpkg into master

strega-nil commented:
/azp run

jgehw commented:
upstream is gone again... please re-run once upstream is fixed

JackBoosY commented:
/azp run

jgehw committed:
Merge branch 'master' of https://github.com/microsoft/vcpkg into master

jgehw committed:
Merge branch 'master' into fix-execute-process-macro2
